### ISSUE TITLE
Issue with Narrowing

### ISSUE THREAD

**Issue by George-Ogden** [python/mypy#21173]
This is a duplicate of #20600, although I cannot reopen it. Here is an example to reproduce. I believe it was introduced in v1.20.0.

```python
from dataclasses import dataclass
from typing import TypeGuard, Self


@dataclass
class ErrorCase:
    x: int

    def __eq__(self, other: object) -> TypeGuard[Self]:
        return (
            type(other)
            is type(self)
        ) and (
                self.x # error: "object" has no attribute "x"
                == other.x
        )
```

The error on line 14 is a false-positive.

**Comment by George-Ogden** [comment]
@hauntsaninja - I think you're the best person to investigate this.

**Comment by hauntsaninja** [comment]
Thanks for the issue! This is related to Self type.

This is not a bad regression, as mypy 1.19 and earlier will complain about this code too, just at a different place.

**Comment by hauntsaninja** [comment]
Fix in https://github.com/python/mypy/pull/21174
